sdk/keyvault/Azure.Security.KeyVault.Shared/src/ChallengeBasedAuthenticationPolicy.cs

```diff
@@ -75,16 +74,18 @@ private async ValueTask ProcessCoreAsync(HttpMessage message, ReadOnlyMemory<Htt
                 message.Request.Content = originalContent;
 
                 // update the cached challenge
-                var challenge = AuthenticationChallenge.GetChallenge(message);
+                challenge = AuthenticationChallenge.GetChallenge(message);
 
-                // if a challenge was returned and its different from the cached _challenge
-                if (challenge != null && !challenge.Equals(_challenge))
+                if (challenge != null)
                 {
-                    // update the cached challenge
-                    _challenge = challenge;
+                    // update the cached challenge if not yet set or different from the current challenge (e.g. moved tenants)
+                    if (_challenge == null || !challenge.Equals(_challenge))
+                    {
+                        _challenge = challenge;
```

Comment:
That won't actually help much since we read `_challenge` in other threads. That - and conflating caching the `_challenge` with `_headerValue` and `_refreshOn` - was the crux of the problem.

Reply:
I'll need to partially rewrite to lock across threads, using an `AutoResetEvent` for example.

Reply:
As we discussed offline, last one wins semantics are acceptable here.

Reply:
It's also worth calling out that parsing the host each time for each request (without caching as a field) would amount to lots of of allocations on the heap because of the string parsing code. That's a separate issue I brought up with @KrzysztofCwalina and have tests with benchmarks at https://github.com/heaths/StringTokenizer.